### PR TITLE
Elektra Web 1.5

### PR DESCRIPTION
**Pull request by omnidan**
- [x] leftovers from #1951 
- [x] array key improvements from usability test
- [x] improved search #1947 
- [x] test with Firefox ESR

@markus2330 please review

---

## Release notes

Many UX improvements from the usability test!

- search completely reworked - it does not act as a filter on already opened keys anymore, and instead searches the whole key database - feedback from the search was also greatly improved (pulsating while searching, glowing blue when done)
- added "abort" buttons to dialogs to revert actions
- added "create array" button to easily create arrays
- removed confirmation dialog before deletion (undo can be used instead)
- created a docker image: `elektra/web`
- small fixes:
  - updated visibility levels
  - removed "done" button in main view
  - fixed issues with the opener click area
  - remove metakeys when they are set to the default value or empty/0

### CODE REVIEW

File: src/tools/web/client/package-lock.json [hunk 1]
```diff
@@ -11324,14 +11324,6 @@
         }
       }
     },
-    "react-tap-event-plugin": {
```
Comment (markus2330) [inline src/tools/web/client/package-lock.json]: Could you please write docu how to udpate this file?

File: src/tools/web/client/src/components/TreeItem/dialogs/SettingsDialog.jsx [hunk 2]
```diff
@@ -99,7 +99,12 @@ export default class SettingsDialog extends Component {
       }
       // persist value to kdb and show notification
       const { timeout } = this.state[key] || {}
-      setMeta(key, value)
+
+      // remove metakey when setting to 0
```
Comment (markus2330) [inline src/tools/web/client/src/components/TreeItem/dialogs/SettingsDialog.jsx]: and empty value?

File: src/tools/web/README.md [hunk 3]
```diff
@@ -153,3 +153,52 @@ authenticate users, e.g. by [username/password auth](https://www.digitalocean.co
...
+  in `client/src/components/TreeItem/dialogs/utils.js`:
+
+```diff
+export const HANDLED_METADATA = [
+  ...,
+  'visibility',
+  'binary',
++ 'check/something',
+]
+```
```
Comment (markus2330) [inline src/tools/web/README.md]: And then, how to do validation or similar?

**Comment by omnidan** [comment]
@markus2330 all tasks from #2032 should be done now, I also tested with Firefox ESR.

you can review already, because only the docker container for the live demo is missing now.

**Comment by markus2330** [comment]
The docker container would make it easier to test (and reproduce) :wink:

**Comment by omnidan** [comment]
@markus2330 I created a Dockerfile (see https://github.com/ElektraInitiative/libelektra/issues/1901#issuecomment-393569029). I also released an image, but it is still 1.4.0 (as the Dockerfile builds from the official git repo).

**Comment by markus2330** [comment]
I am okay with merging this PR so that creating the Docker image is easier. If everything is finished, please set "ready to merge".

File: src/tools/web/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,34 @@
+FROM ubuntu:16.04
```
Comment (markus2330) [inline src/tools/web/Dockerfile]: Any reason why this file is not in scripts/docker/?

**Comment by omnidan** [comment]
@markus2330 I already created a `1.5.0-beta` image, no worries. you can install it via:

```
docker run -d -it -p 33333:33333 -p 33334:33334 elektra/web:1.5.0-beta
```

**Comment by ingwinlu** [comment]
why the two ports? you are missing `EXPOSE` entries in the Dockerfile. You might also be interested in removing some layers to reduce image size and speed up builds.

Additionally you don't have a PID1 that controls signals sent to the container.

**Comment by markus2330** [comment]
@omnidan Thank you, works for me. As said, mountpoints would make the image more interesting. (Maybe mount /tmp/hosts to not destroy anything.) What about providing a start-web script instead of creating it only for the docker image? And release notes are missing! You are welcome to highlight your changes.

@ingwinlu Thank you for reviewing the Dockerfile!

**Comment by markus2330** [comment]
@omnidan small things: "edit value" does not automatically focus to the edit field. Please also check if tabs work as wanted (they also cycle in the background of the dialog). And numbering of arrays sometimes does not work (if I clicked somewhere else before.) Should I create an issue for this?

**Comment by omnidan** [comment]
@ingwinlu thanks for the hints! I added the two ports via EXPOSE. One port is for elektrad (33333), the other for webd/client (33334). I also tried reducing the image size by following instructions from: https://blog.florianlopes.io/5-tips-to-reduce-docker-image-size/

I'm not sure what you mean by "you don't have a PID 1" - I found out that nodejs is not supposed to run as PID 1 in docker: https://www.elastic.io/nodejs-as-pid-1-under-docker-images/

**Comment by omnidan** [comment]
@markus2330 

> "edit value" does not automatically focus to the edit field. 

thanks for the hint - I fixed this!

> Please also check if tabs work as wanted (they also cycle in the background of the dialog).

unfortunately trapping the focus in the dialog does not work with the current version of the UI library I use (MUI). I tried using an external library to achieve this, but with the external library I can only trap focus within the dialog contents, which excludes the dialog buttons. so when I enable the focus trap, it is not possible to navigate to the dialog buttons anymore.

focus trapping dialogs are implemented in the latest beta version of MUI, however: https://github.com/mui-org/material-ui/issues/4384 - should I make an issue about things that can be fixed by upgrading to MUI 1.0 when it comes out (selecting text when pressing the copy button, trapping focus in dialogs)?

> And numbering of arrays sometimes does not work (if I clicked somewhere else before.)

Where did you click? I'm not sure how to reproduce this.

**Comment by omnidan** [comment]
@markus2330 

> Thank you, works for me. As said, mountpoints would make the image more interesting. (Maybe mount /tmp/hosts to not destroy anything.)

We could push an example kdb export to a git repo and then pull that to build the docker image for the live demo (I would create a separate image for that though).

> And release notes are missing! You are welcome to highlight your changes.

I added them to the initial post of this PR!

**Comment by markus2330** [comment]
Thank you look great.

But please describe changes in changelog (news). (your text in this PR is quite good, but it should be in doc/news/_preparation_next_release.md)

You can also finish (some) tasks of #2037 here.

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Please add some info to the release notes.

**Comment by markus2330** [comment]
Thank you. Seems like you need to rebase.

**Comment by omnidan** [comment]
@markus2330 done!

**Comment by markus2330** [comment]
Thank you! Could you please rebase to get rid of the old commits.

@ingwinlu Can you take a look at the docker file?

**Comment by omnidan** [comment]
@markus2330 I already rebased, what do you mean "get rid of the old commits"?

**Comment by markus2330** [comment]
The github view shows 250+ commits. Something seems to be wrong here.

**Comment by markus2330** [comment]
And maybe it is a good idea to split the docker files and the WebUI changes.

**Comment by omnidan** [comment]
> The github view shows 250+ commits. Something seems to be wrong here.

@markus2330 that is because I rebased and it added all commits from upstream master to the branch, they should not be duplicated after merging.

**Comment by omnidan** [comment]
> And maybe it is a good idea to split the docker files and the WebUI changes.

I will remove them from this PR for now and create a new PR with the updated docker files tomorrow.

**Comment by markus2330** [comment]
There is still something very wrong in this PR. There are 250+ commits.

**Comment by omnidan** [comment]
@markus2330 as I said, that is a result of rebasing, it happened in the previous PR too. Look at the "Files changed" tab, it does not actually contain the changes of the commits.

**Comment by omnidan** [comment]
see #2099

**Comment by ingwinlu** [comment]
@omnidan out of curiosity: how do you rebase?

**Comment by markus2330** [comment]
> it happened in the previous PR too.

This does not say it was correct. It also messed up the history back then.

The problem is that the common commit is old (26.05) and that you rebased something that already contained a merge commit.

A rebase should be very simple in your case, nobody else modified the directories you are working in.

You can also create a new PR if it is easier for you.

**Comment by omnidan** [comment]
@ingwinlu https://help.github.com/articles/syncing-a-fork/ and then `git rebase master`

**Comment by markus2330** [comment]
I do not think you should mix these two.

You can either merge with `-ff` or even `-ff-only` or do a pull with `--rebase`.

**Comment by ingwinlu** [comment]
Make sure upstream is really pointing at the correct repository via `git remote -v`.
Your process seems fine to me otherwise.

**Comment by omnidan** [comment]
@ingwinlu I read that it might be the way I created the branch, not the way I rebased: https://stackoverflow.com/a/40413455

**Pull request closed without merging**